File: bench/py/18_foreach_array.py
# for-each over a list
import sys
scale = int(sys.argv[1]) if len(sys.argv) > 1 else 1

a = list(range(1000000))

R = 3 * scale
s = 0

for k in range(R):
    for e in a:
        s += e
    s = s % 1000000007

print("result:", s)

File: bench/cpp/18_foreach_array.cpp
/* SPDX-License-Identifier: BSD-2-Clause */
/* Faithful C++ of bench/my/18_foreach_array.my: foreach over a 1M-element array,
 * repeated R times, accumulating with a mod after each full pass. */
#include "bench.h"

int main(int argc, char **argv)
{
    long scale = bench_scale(argc, argv);

    std::vector<long> a(1000000);       /* range(1000000) */
    for (long i = 0; i < 1000000; i++)
        a[i] = i;

    long R = 3 * scale;
    long s = 0;

    for (long k = 0; k < R; k++) {
        /* MyLang re-scans the whole array every outer pass; block the compiler
         * from hoisting the loop-invariant reduction out of the outer loop. */
        bench_sink_ptr(a.data());
        for (long e : a)
            s += e;
        s = s % 1000000007;
    }

    printf("result: %ld\n", s);
    return 0;
}
